File: golang/155.min-stack.go
type MinStack struct {
	st [][2]int // 0: value, 1: current min
}

func Constructor() MinStack {
	return MinStack{
		st: [][2]int{},
	}
}

func (this *MinStack) Push(val int) {
	var mi int
	n := len(this.st)
	if n == 0 {
		mi = val
	} else {
		mi = this.st[n-1][1]
	}

	mi = min(mi, val)
	this.st = append(this.st, [2]int{val, mi})
}

func (this *MinStack) Pop() {
	this.st = this.st[:len(this.st)-1]
}

func (this *MinStack) Top() int {
	return this.st[len(this.st)-1][0]
}

func (this *MinStack) GetMin() int {
	return this.st[len(this.st)-1][1]
}

File: python/155.min-stack.py
class MinStack:

    def __init__(self):
        self._stack = []
        self._min = None

    def push(self, val: int) -> None:
        if self._min is None or self._min > val:
            self._min = val
        self._stack.append(val)

    def pop(self) -> None:
        val = self._stack.pop()
        if len(self._stack):
            self._min = min(self._stack)
        else:
            self._min = None
        return val

    def top(self) -> int:
        if len(self._stack):
            return self._stack[-1]
        return 0

    def getMin(self) -> int:
        return self._min



# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()
